### PR TITLE
gh-104050: Argument clinic: improve typing around adding C converters

### PR DESCRIPTION
**Pull request by AlexWaygood**
Some of this looks a _little_ complicated at first glance, but it's hopefully not too hard to grok

* Issue: gh-104050

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -2647,10 +2648,12 @@ def __getattribute__(self, name: str) -> Any:
...
 
 
+CConverterClassT = TypeVar("CConverterClassT", bound=type["CConverter"])
+
 def add_c_converter(
-        f: type[CConverter],
+        f: CConverterClassT,
         name: str | None = None
-) -> type[CConverter]:
+) -> CConverterClassT:
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:2656]: By using a `TypeVar` here (and with the other functions in this PR), we can make the annotation much more expressive.

The annotation currently says "you put in an object `f` that's some subtype of `type[CConverter]`, and you get out an object that's some subtype of `type[CConverter]`.

The new annotation, by contrast, says, "you put in an object `f`, and you get out an object that has the same type as whatever the type of `f` is, _as long as_ the type of `f` is a subtype of `type[CConverter]`.

Comment (AlexWaygood) [reply Tools/clinic/clinic.py:2656]: As an example, with the following snippet:

```py
x = add_c_converter(bool_converter)
```

Mypy would have inferred the type of `x` to be `type[CConverter]` with the current annotation. But with the new annotation, mypy should be able to infer that the type of `x` is `bool_converter`.

Comment (erlend-aasland) [reply Tools/clinic/clinic.py:2656]: Neat!

**Review by erlend-aasland — APPROVED** [review]
Thanks!

**Pull request merged**